Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1918,66 +1948,161 @@ void CGContextSetPatternPhase(CGContextRef context, CGSize phase) {
+            }
+            return S_OK;
+        }));
     } else {
-        // Using device space here gives us finer-grained control over the transform.
-        RETURN_IF_FAILED(
-            Draw(_kCGCoordinateModeDeviceSpace,
-                 nullptr,
-                 [&state, &glyphRun, &deviceTransform, &textTransform](CGContextRef context, ID2D1DeviceContext* deviceContext) {
-                     CGAffineTransform runningGlobalTransform = deviceTransform;
-                     // Text scaling and rotation apply to each glyph relative to its origin, so we must draw each glyph
-                     // transformed
-                     // independently
-                     DWRITE_GLYPH_RUN individualGlyphRun{ glyphRun->fontFace,
-                                                          glyphRun->fontEmSize,
-                                                          1, // Since this is glyph by glyph, glyphCount is one
-                                                          glyphRun->glyphIndices,
-                                                          nullptr,
-                                                          nullptr,
-                                                          glyphRun->isSideways,
-                                                          glyphRun->bidiLevel };
-                     D2D1_POINT_2F origin{ 0, 0 };
-                     // Iterate through every glyph by incrementing pointer in glyphIndices array
-                     for (uint32_t i = 0; i < glyphRun->glyphCount; ++i, ++(individualGlyphRun.glyphIndices)) {
-                         CGAffineTransform finalTextTransform = CGAffineTransformConcat(textTransform, runningGlobalTransform);
-                         deviceContext->SetTransform(__CGAffineTransformToD2D_F(finalTextTransform));
-                         deviceContext->DrawGlyphRun(origin, &individualGlyphRun, state.fillBrush.Get(), DWRITE_MEASURING_MODE_NATURAL);
-
-                         // Uses glyphAdvances to move each glyph
-                         runningGlobalTransform = CGAffineTransformTranslate(runningGlobalTransform, glyphRun->glyphAdvances[i], 0);
-                     }
-                     return S_OK;
-                 }));
+        // Get the linear transformation that is the inverse of the text transform
+        // We transform the positions of each glyph by the inverse so they will be positioned correctly
+        // While still being transformed per glyph by the text transformation
+        CGAffineTransform invertedTextTransformation = CGAffineTransformInvert(
+            CGAffineTransformScale(CGAffineTransformMake(textTransform.a, textTransform.b, textTransform.c, textTransform.d, 0, 0), 1, -1));
+
+        // Vectors which hold the positions of each glyph that we give to DWRITE for drawing
+        std::vector<std::vector<DWRITE_GLYPH_OFFSET>> positionsVectors(runsCount);
```

Comment:
thinking on it more it might actually be that you can't avoid constructing the glyph runs below, in which case you could construct them up here, and use their glyphOffsets field to store a ptr to either original glyphOffsets or transformed glyphOffsets (positionsVectors and glyphRuns sizes would become intentionally out of sync)

---
In reply to: [100935594](https://github.com/Microsoft/WinObjC/pull/1933#discussion_r100935594) [](ancestors = 100935594)